{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "83e57e3a-e05b-434c-935a-0ff2783b0e5f",
   "metadata": {},
   "source": [
    "# Table of Contents\n",
    "- [Processing Functions](#Process-functions)\n",
    "- [Import Data](#import-data)\n",
    "  - [Histograms](#Histograms)\n",
    "  - [Counts files](#counts)\n",
    "- [Processing](#processing)\n",
    "- [Plotting](#plotting)\n",
    "  - [HCT116-Rpb1 degron treatment with TRP/IAA, Biological Replicate 1](#LP-88-HCT116-Rpb1-degron-Biological-Replicate-1)\n",
    "  - [HCT116-Rpb1 degron treatment with TRP/IAA, Biological Replicate 2](#LP-91-HCT116-Rpb1-degron-Biological-Replicate-2)\n",
    "  - [HCT116-Rpb1 degron treatment with TRP/IAA, Biological Replicates 3 and 4](#LP-95-HCT116-Rpb1-degron-Biological-Replicates-3-and-4)\n",
    "- [Spike-in Normalize](#spike-in-normalize)\n",
    "  - [dual norm with confidence intervals](#dual-norm-with-confidence-intervals)\n",
    "\n",
    "---"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "56747a7b-b206-479f-987c-86bc644f0d6c",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "── \u001b[1mAttaching core tidyverse packages\u001b[22m ──────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────── tidyverse 2.0.0 ──\n",
      "\u001b[32m✔\u001b[39m \u001b[34mdplyr    \u001b[39m 1.1.4     \u001b[32m✔\u001b[39m \u001b[34mreadr    \u001b[39m 2.1.5\n",
      "\u001b[32m✔\u001b[39m \u001b[34mforcats  \u001b[39m 1.0.1     \u001b[32m✔\u001b[39m \u001b[34mstringr  \u001b[39m 1.5.2\n",
      "\u001b[32m✔\u001b[39m \u001b[34mggplot2  \u001b[39m 4.0.0     \u001b[32m✔\u001b[39m \u001b[34mtibble   \u001b[39m 3.3.0\n",
      "\u001b[32m✔\u001b[39m \u001b[34mlubridate\u001b[39m 1.9.4     \u001b[32m✔\u001b[39m \u001b[34mtidyr    \u001b[39m 1.3.1\n",
      "\u001b[32m✔\u001b[39m \u001b[34mpurrr    \u001b[39m 1.1.0     \n",
      "── \u001b[1mConflicts\u001b[22m ────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────── tidyverse_conflicts() ──\n",
      "\u001b[31m✖\u001b[39m \u001b[34mdplyr\u001b[39m::\u001b[32mfilter()\u001b[39m masks \u001b[34mstats\u001b[39m::filter()\n",
      "\u001b[31m✖\u001b[39m \u001b[34mdplyr\u001b[39m::\u001b[32mlag()\u001b[39m    masks \u001b[34mstats\u001b[39m::lag()\n",
      "\u001b[36mℹ\u001b[39m Use the conflicted package (\u001b[3m\u001b[34m<http://conflicted.r-lib.org/>\u001b[39m\u001b[23m) to force all conflicts to become errors\n"
     ]
    }
   ],
   "source": [
    "library(tidyverse)\n",
    "theme_set(theme_classic())\n",
    "library(DescTools)\n",
    "library(RColorBrewer)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "8aa10b64-7893-4b24-a63e-1a11d39559d4",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "id": "fd59e0f9-9379-4689-b863-93a312ac420c",
   "metadata": {},
   "source": [
    "## Processing HOMER Histograms\n",
    "\n",
    "The functions: `process_histograms_dm6` `process_histograms_sac3` and `process_histograms_hg38` work with histogram files generated with HOMER annotatePeaks with -hist option. More details are available [on the HOMER tutorial website](http://homer.ucsd.edu/homer/ngs/quantification.html).\n",
    "\n",
    "Making the histogram: \n",
    "\n",
    "`annotatePeaks.pl [region_to_center_on] [genome_version] -size [size_to_plot] -hist [bp_resolution] -d [tag directories] > histogram_regions_genome_datasource.txt`\n",
    "\n",
    "Example for a TSS histogram in the hg38 genome: \n",
    "\n",
    "`annotatePeaks.pl tss hg38 -size 4000 -hist 25 -d sample1-tagdir sample2-tagdir input1-tagdir input2-tagdir > hist_tss_hg38_samples12.txt`\n",
    "\n",
    "Input file: histogram_regions_genome_datasource.txt \n",
    "\n",
    "Output: dataframe in tidy (long) format, with additional columns describing the metadata originally in each sample name:\n",
    "- Cell Type\n",
    "- Treatment\n",
    "- Timepoint\n",
    "- Biological Replicate\n",
    "- Antibody (usually I have multiple IP samples from each biological sample/replicate)\n",
    "- Technical Replicate (separate library preps from the same biological sample/replicate)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "213aa84a-faff-42b3-85ec-5e15ff560421",
   "metadata": {},
   "outputs": [],
   "source": [
    "process_histograms_dm6 <- function(x, .x) {\n",
    "    colnames(x)[1] <- \"Distance_from_tss\"\n",
    "    x <- x %>% \n",
    "      ### first two rename_with uses are specific to my sample names\n",
    "      # get rid of suffix to samples, leaving only sample name.Coverage\n",
    "    rename_with(~ gsub(\"\\\\.+\\\\.Coverage\", \".Coverage\", .x), contains(\"tagdir\")) %>% \n",
    "    rename_with(~ gsub(\"_H3K9ac\", \"_1_H3K9ac\", .x), contains(\"Hela\")) %>%\n",
    "    rename_with(~ gsub(\"_input\", \"_1_input\", .x), contains(\"sync\")) %>%\n",
    "      # get rid of any prefix before cell type (ex: directory path)\n",
    "    rename_with(~ gsub(\".+\\\\_HCT\", \"HCT\", .x), contains(\"HCT\")) %>%\n",
    "    # get rid of any suffix (ex: S2, S4)\n",
    "    rename_with(~ gsub(\"_S[0-9]+\", \"\", .x), contains(\"Hela|HCT\")) %>%\n",
    "    # add techreps to inputs\n",
    "    rename_with(~ gsub(\"input_R1_001.\", \"input_1.\", .x), contains(\"HCT\")) %>%\n",
    "     rename_with(~ gsub(\"_R1_001\", \"\", .x), contains(\"Hela\")) %>%\n",
    "    rename_with(~ gsub(\"input.\", \"input_1.\", .x), contains(\"HCT\")) %>%\n",
    "      ### below two uses of rename_with are general to HOMER output\n",
    "    rename_with(~ gsub(\"\\\\.[[:digit:]]$\", \"_minus\", .x), contains(\"Tags\")) %>% \n",
    "    rename_with(~ gsub(\"\\\\.\\\\.\\\\.\", \"_\", .x), contains(\"Tags\"))\n",
    "    \n",
    "    # select coverage columns only\n",
    "    xcov <- x %>% select(contains(\"Coverage\"))\n",
    "    xcov$Distance_from_tss <- x$Distance_from_tss\n",
    "    \n",
    "    xcovlong <- \n",
    "    xcov %>% pivot_longer(\n",
    "      cols = -\"Distance_from_tss\", \n",
    "      names_to = \"Sample\", \n",
    "      values_to = \"Coverage\")\n",
    "    xcovlong <- xcovlong %>%\n",
    "  mutate(\n",
    "    cell = str_match(Sample, '([^_]+)(?:_[^_]+){5}$')[,2],\n",
    "    treatment = str_match(Sample, '([^_]+)(?:_[^_]+){4}$')[,2],\n",
    "    timepoint = str_remove(str_match(Sample, '([^_]+)(?:_[^_]+){3}$')[,2], \"TSA\"),\n",
    "    biorep = str_match(Sample, '([^_]+)(?:_[^_]+){2}$')[,2],\n",
    "    antibody = str_match(Sample, '([^_]+)(?:_[^_]+){1}$')[,2],\n",
    "    techrep = str_match(Sample, '([^_]+)(?:_[^_]+){0}$')[,2]\n",
    "  )\n",
    "  }\n",
    "\n",
    "process_histograms_sac3 <- function(x, .x) {\n",
    "    colnames(x)[1] <- \"Distance_from_tss\"\n",
    "    x <- x %>% \n",
    "      ### first two rename_with uses are specific to my sample names\n",
    "      # get rid of suffix to samples, leaving only sample name.Coverage\n",
    "    rename_with(~ gsub(\".concat.+Coverage\", \".Coverage\", .x), contains(\"tagdir\")) %>% \n",
    "  #  rename_with(~ gsub(\"_H3K9ac\", \"_1_H3K9ac\", .x), contains(\"Hela\")) %>%\n",
    " #   rename_with(~ gsub(\"_input\", \"_1_input\", .x), contains(\"sync\")) %>%\n",
    "      # get rid of any prefix before cell type (ex: directory path)\n",
    "    rename_with(~ gsub(\".+\\\\_HCT\", \"HCT\", .x), contains(\"HCT\")) %>%\n",
    "    # get rid of any suffix (ex: S2, S4)\n",
    "    rename_with(~ gsub(\"_S[0-9]+\", \"\", .x), contains(\"Hela|HCT\")) %>%\n",
    "    # add techreps to inputs\n",
    "    rename_with(~ gsub(\"input.\", \"input_1.\", .x), contains(\"HCT\")) %>%\n",
    "    rename_with(~ gsub(\"input_R1_001.\", \"input_1.\", .x), contains(\"HCT\")) %>%\n",
    "     rename_with(~ gsub(\"_R1_001\", \"\", .x), contains(\"Hela\")) %>%\n",
    "      ### below two uses of rename_with are general to HOMER output\n",
    "    rename_with(~ gsub(\"\\\\.[[:digit:]]$\", \"_minus\", .x), contains(\"Tags\")) %>% \n",
    "    rename_with(~ gsub(\"\\\\.\\\\.\\\\.\", \"_\", .x), contains(\"Tags\"))\n",
    "    \n",
    "    # select coverage columns only\n",
    "    xcov <- x %>% select(contains(\"Coverage\"))\n",
    "    xcov$Distance_from_tss <- x$Distance_from_tss\n",
    "    \n",
    "    xcovlong <- \n",
    "    xcov %>% pivot_longer(\n",
    "      cols = -\"Distance_from_tss\", \n",
    "      names_to = \"Sample\", \n",
    "      values_to = \"Coverage\")\n",
    "    xcovlong <- xcovlong %>%\n",
    "  mutate(\n",
    "    cell = str_match(Sample, '([^_]+)(?:_[^_]+){5}$')[,2],\n",
    "    treatment = str_match(Sample, '([^_]+)(?:_[^_]+){4}$')[,2],\n",
    "    timepoint = str_remove(str_match(Sample, '([^_]+)(?:_[^_]+){3}$')[,2], \"TSA\"),\n",
    "    biorep = str_match(Sample, '([^_]+)(?:_[^_]+){2}$')[,2],\n",
    "    antibody = str_match(Sample, '([^_]+)(?:_[^_]+){1}$')[,2],\n",
    "    techrep = str_match(Sample, '([^_]+)(?:_[^_]+){0}$')[,2]\n",
    "  )\n",
    "  }\n",
    "\n",
    "process_histograms_hg38 <- function(x, .x) {\n",
    "    colnames(x)[1] <- \"Distance_from_tss\"\n",
    "    x <- x %>% \n",
    "      ### first two rename_with uses are specific to my sample names\n",
    "      # get rid of suffix to samples, leaving only sample name.Coverage\n",
    "    rename_with(~ gsub(\".concat.hg38.tagdir.Coverage\", \".Coverage\", .x), contains(\"HCT\")) %>% \n",
    "    rename_with(~ gsub(\".concat.dedup.hg38.filtered.tagdir\", \"\", .x), contains(\"tagdir\")) %>% \n",
    "      # get rid of any prefix before cell type (ex: directory path)\n",
    "  #  rename_with(~ gsub(\"_H3K9ac\", \"_1_H3K9ac\", .x), contains(\"Hela\")) %>%\n",
    "    rename_with(~ gsub(\"_input\", \"_1_input\", .x), contains(\"sync\")) %>%\n",
    "    rename_with(~ gsub(\".+\\\\_HCT\", \"HCT\", .x), contains(\"HCT\")) %>% \n",
    "     # get rid of any suffix (ex: S2, S4)\n",
    "    rename_with(~ gsub(\"_S[0-9]+\", \"\", .x), contains(\"Hela|HCT\")) %>%\n",
    "    # add techreps to inputs\n",
    "#    rename_with(~ gsub(\"input.\", \"input_1.\", .x), contains(\"HCT\")) %>%\n",
    "    rename_with(~ gsub(\"input_R1_001.\", \"input_1.\", .x), contains(\"HCT\")) %>%\n",
    "     rename_with(~ gsub(\"_R1_001\", \"\", .x), contains(\"Hela\")) %>%\n",
    "      ### below two uses of rename_with are general to HOMER output\n",
    "    rename_with(~ gsub(\"\\\\.[[:digit:]]$\", \"_minus\", .x), contains(\"Tags\")) %>% \n",
    "    rename_with(~ gsub(\"\\\\.\\\\.\\\\.\", \"_\", .x), contains(\"Tags\"))\n",
    "    \n",
    "    # select coverage columns only\n",
    "    xcov <- x %>% select(contains(\"Coverage\"))\n",
    "    xcov$Distance_from_tss <- x$Distance_from_tss\n",
    "    \n",
    "    xcovlong <- \n",
    "    xcov %>% pivot_longer(\n",
    "      cols = -\"Distance_from_tss\", \n",
    "      names_to = \"Sample\", \n",
    "      values_to = \"Coverage\")\n",
    "    \n",
    "    xcovlong <- xcovlong %>%\n",
    "  mutate(\n",
    "    cell = str_match(Sample, '([^_]+)(?:_[^_]+){5}$')[,2],\n",
    "    treatment = str_match(Sample, '([^_]+)(?:_[^_]+){4}$')[,2],\n",
    "    timepoint = str_remove(str_match(Sample, '([^_]+)(?:_[^_]+){3}$')[,2], \"TSA\"),\n",
    "    biorep = str_match(Sample, '([^_]+)(?:_[^_]+){2}$')[,2],\n",
    "    antibody = str_match(Sample, '([^_]+)(?:_[^_]+){1}$')[,2],\n",
    "    techrep = str_match(Sample, '([^_]+)(?:_[^_]+){0}$')[,2]\n",
    "  )\n",
    "  }"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "9bb3d576-84cd-487d-a92c-e96f54b21911",
   "metadata": {},
   "source": [
    "## Processing HOMER counts files\n",
    "\n",
    "The function `process_counts_annotpeaks()` takes HOMER annotatePeaks files generated without the -hist option. Instead of generating average Coverage across all regions specified, the resulting annotated counts file is a matrix, with rows for each region. The first 19 columns are metadata describing each region, the remaining columns are the Read Normalized Tag Counts for each sample at each region.\n",
    "\n",
    "Relevant Metadata Columns: \n",
    "1. **PeakID**\n",
    "2. **Chr**\n",
    "3. **Start**\n",
    "4. **End**\n",
    "5. Strand - strand that nearest annotated gene/TSS is located on, not necessarily strand of your data! Remember ChIP-seq/ATAC-seq is unstranded by definition. \n",
    "6. Not Used - ignore\n",
    "7. Focus.Ratio.Region.Size\n",
    "8. **Annotation** - if annotating on TSS regions, all will be promoter-TSS\n",
    "9. Detailed Annotation\n",
    "10. **Distance to TSS** - if annotating on TSS regions, all will be 0."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "91fd01e4-0b04-4917-a48c-dffb82a9df33",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "id": "1256dfaf-7726-44ca-b0b9-d471fd7c07d0",
   "metadata": {},
   "source": [
    "# import data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "d35ac26e-ef7d-42d9-b764-53c7eff0114b",
   "metadata": {},
   "outputs": [],
   "source": [
    "# LP 88 - HCT116-Rpb1 degron biological replicate 1\n",
    "hist_tss_hg38_LP88 <- read.delim(\"/home/lahodge/data/experiments/LP_88_combined_fastqs/hg38_data/hg38_tagdirs/hist_tss_hg38_LP88.txt\")\n",
    "hist_K27ac_hg38_LP88 <- read.delim(\"/home/lahodge/data/experiments/LP_88_combined_fastqs/hg38_data/hg38_tagdirs/hist_H3K27acpeaks_hg38_LP88.txt\")\n",
    "hist_over100kb_hg38_LP88 <- read.delim(\"/home/lahodge/data/experiments/LP_88_combined_fastqs/hg38_data/hg38_tagdirs/hist_over100kbgenes_hg38_LP88.txt\")\n",
    "\n",
    "# LP 91 - HCT116-Rpb1 degron biological replicate 2, technical replicate 1/2\n",
    "hist_shortrna_hg38_LP91 <- read.delim(\"/home/lahodge/data/experiments/LP_91/hg38_data/hg38_tagdirs/hist_5to20kb_hg38_LP91.txt\")\n",
    "hist_tss_hg38_LP91 <- read.delim(\"/home/lahodge/data/experiments/LP_91/hg38_data/hg38_tagdirs/hist_tss_hg38_LP91.txt\")\n",
    "hist_rna_hg38_LP91 <- read.delim(\"/home/lahodge/data/experiments/LP_91/hg38_data/hg38_tagdirs/hist_rna_hg38_LP91.txt\")\n",
    "\n",
    "# LP 95 - HCT116-Rpb1 degron biological replicates 3 and 4"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "17b033e4-fd6e-4503-a39f-b2e998193524",
   "metadata": {},
   "outputs": [],
   "source": [
    "hist_tss_hg38_LP88_long <- process_histograms_dm6(hist_tss_hg38_LP88)\n",
    "hist_over100kb_hg38_LP88_long <- process_histograms_hg38(hist_over100kb_hg38_LP88)\n",
    "\n",
    "hist_shortrna_hg38_LP91_long <- process_histograms_hg38(hist_shortrna_hg38_LP91)\n",
    "hist_tss_hg38_LP91_long <- process_histograms_hg38(hist_tss_hg38_LP91)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "518b0d1e-e9ed-463c-b936-49ef1ad68c54",
   "metadata": {},
   "source": [
    "## LP 88 HCT116 Rpb1 degron Biological Replicate 1"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "202277cf-5584-4f42-b313-47e3581fbdcc",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "id": "e3c36597-395d-426a-b3a7-5b0c6f744ecd",
   "metadata": {},
   "source": [
    "### *H. sapiens* data\n",
    "\n",
    "We quantify signal at regions biologically relevant for each antibody:\n",
    "- H3K27ac: at TSSs or H3K27ac peaks\n",
    "- H3K4me3: at TSSs or H3K4me3 peaks\n",
    "- H3K4me1: at H3K4me1 peaks\n",
    "- H3K36me3: at gene bodies or H3K36me3 peaks\n",
    "- Rpb1: at TSSs or Rpb1 peaks"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "eae71b87-72e1-4b3f-b73d-af26c8246180",
   "metadata": {},
   "source": [
    "## Scatterplot comparing treatment changes"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "34e32a63-101b-415d-b85e-e0416e008a98",
   "metadata": {},
   "outputs": [],
   "source": [
    "some_function2_FC <- function(df, quantile, xnum, xdenom, ynum, ydenom, regions, colorbrewer) {\n",
    "  # extract sample conditions from xvariable name\n",
    "    cell_type <- str_match(xnum, '([^_]+)(?:_[^_]+){5}$')[,2]\n",
    "    ### x axis variables\n",
    "    treatment_xnum <- str_match(xnum, '([^_]+)(?:_[^_]+){4}$')[,2]\n",
    "    timepoint_xnum <- str_match(xnum, '([^_]+)(?:_[^_]+){3}$')[,2]\n",
    "    treatment_xdenom <- str_match(xdenom, '([^_]+)(?:_[^_]+){4}$')[,2]\n",
    "    timepoint_xdenom <- str_match(xdenom, '([^_]+)(?:_[^_]+){3}$')[,2]\n",
    "    biorep_xnum <- str_match(xnum, '([^_]+)(?:_[^_]+){2}$')[,2]\n",
    "    biorep_xdenom <- str_match(xdenom, '([^_]+)(?:_[^_]+){2}$')[,2]\n",
    "    antibody_xnum <- str_match(xnum, '([^_]+)(?:_[^_]+){1}$')[,2]\n",
    "    antibody_xdenom <- str_match(xdenom, '([^_]+)(?:_[^_]+){1}$')[,2]\n",
    "    xnumcondition <- paste(antibody_xnum, treatment_xnum, timepoint_xnum)\n",
    "    xdenomcondition <- paste(antibody_xdenom, treatment_xdenom, timepoint_xdenom)\n",
    "    ### y axis variables\n",
    "    treatment_ynum <- str_match(ynum, '([^_]+)(?:_[^_]+){4}$')[,2]\n",
    "    timepoint_ynum <- str_match(ynum, '([^_]+)(?:_[^_]+){3}$')[,2]\n",
    "    treatment_ydenom <- str_match(ydenom, '([^_]+)(?:_[^_]+){4}$')[,2]\n",
    "    timepoint_ydenom <- str_match(ydenom, '([^_]+)(?:_[^_]+){3}$')[,2]\n",
    "    biorep_ynum <- str_match(ynum, '([^_]+)(?:_[^_]+){2}$')[,2]\n",
    "    biorep_ydenom <- str_match(ydenom, '([^_]+)(?:_[^_]+){2}$')[,2]\n",
    "    antibody_ynum <- str_match(ynum, '([^_]+)(?:_[^_]+){1}$')[,2]\n",
    "    antibody_ydenom <- str_match(ydenom, '([^_]+)(?:_[^_]+){1}$')[,2]\n",
    "    ynumcondition <- paste(antibody_ynum, treatment_ynum, timepoint_ynum)\n",
    "    ydenomcondition <- paste(antibody_ydenom, treatment_ydenom, timepoint_ydenom)\n",
    "  # make sure regions and colorbrewer varaiables are interpretable\n",
    "  regions <- rlang::ensym(regions)\n",
    "  colorbrewer <- rlang::ensym(colorbrewer)\n",
    "  \n",
    "  # need to convert quantile column to a factor (not 1-4 numeric) for plotting colors to work\n",
    "#  df[quantile] <- as.factor(df[, quantile]) \n",
    "  num_colors <- length(unique(df[[\"quantile\"]]))\n",
    "  \n",
    "  # make the plot\n",
    "ggplot(data = df) + \n",
    "    aes(x = .data[[xnum]]/.data[[xdenom]], y = (.data[[ynum]])/(.data[[ydenom]]), \n",
    "        color = as.factor(quantile)) + \n",
    "    scale_x_log10() + scale_y_log10() + \n",
    "    geom_hline(yintercept = 1) + geom_vline(xintercept = 1) + \n",
    "    geom_point(alpha = 0.05, stroke = NA) + \n",
    "    labs(\n",
    "      title = paste(\"comparing\", xnumcondition, \"and\", ynumcondition, \"changes\"), \n",
    "      subtitle = paste(\"in\", cell_type, \"at\", regions),\n",
    "      x = paste(treatment_xnum, timepoint_xnum, \"/\", treatment_xdenom, timepoint_xdenom), \n",
    "      y = paste(treatment_ynum, timepoint_ynum, \"/\", treatment_ydenom, timepoint_ydenom)) +\n",
    "  scale_color_manual(\n",
    "      values = colorRampPalette(brewer.pal(9, paste0(colorbrewer)))(7)[4:7]) + \n",
    "  theme(legend.position = \"none\")\n",
    "}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "3a7ada6d-0b86-49de-a7b7-3ca4a44f21ae",
   "metadata": {},
   "outputs": [],
   "source": [
    "options(repr.plot.width = 7, repr.plot.height = 6.5)\n",
    "\n",
    "some_function2_FC(counts_K27_hg38_dual %>%\n",
    "                    filter(HCT116_DMSO_0hr_1_H3K27ac_1.Tag > \n",
    "                             mean(HCT116_DMSO_0hr_1_H3K27ac_1.Tag)), quantile, \n",
    "                  xnum = \"HCT116_TRP_4hr_1_H3K27ac_1.Tag\", \n",
    "                  xdenom = \"HCT116_DMSO_0hr_1_H3K27ac_1.Tag\", \n",
    "                  ynum = \"HCT116_IAA_4hr_1_H3K27ac_1.Tag\", \n",
    "                  ydenom = \"HCT116_DMSO_0hr_1_H3K27ac_1.Tag\", \n",
    "                  regions = \"shared H3K27ac peaks\", \n",
    "                  colorbrewer = \"RdYlBu\") +\n",
    "  aes(color = log(abs(Distance.to.TSS))) + \n",
    "  geom_point(alpha = 0.6, stroke = NA, size = 2) + \n",
    "scale_color_viridis() +\n",
    "  theme(legend.position = c(0.82, 0.22), \n",
    "       axis.text = element_text(size=20), \n",
    "        axis.title=element_text(size=20), \n",
    "       legend.text = element_text(size=15),\n",
    "       legend.title = element_text(size=18)) "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "261bd3ca-f7d7-4b3e-9068-346250862a95",
   "metadata": {},
   "outputs": [],
   "source": [
    "counts_K27_hg38_dual$Annotation <- gsub(\"\\\\(.*\", \"\", as.character(counts_K27_hg38_dual$Annotation))\n",
    "counts_K27_hg38_dual$Annotation <- gsub(\" \", \"\", as.character(counts_K27_hg38_dual$Annotation))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "53bf9ea5-2e49-4b8d-8be9-148f5bc5fa9d",
   "metadata": {},
   "outputs": [],
   "source": [
    "counts_K27_hg38_dual_clean <- counts_K27_hg38_dual %>%\n",
    "  mutate(\n",
    "    Annotation_Clean = case_when(\n",
    "      Annotation %in% c(\"intron\", \"exon\", \"5'UTR\", \"3'UTR\", \"TTS\") ~ \"genic_regions\",\n",
    "      Annotation %in% c(\"Intergenic\", \"non-coding\") ~ \"distal_nongenic_regions\",\n",
    "      Annotation == \"promoter-TSS\" ~ \"promoter-TSS\",\n",
    "      TRUE ~ Annotation  # fallback: keep original if not matched\n",
    "    )\n",
    "  )"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d6f1ccfa-e198-4f51-91fd-da8367cd48e5",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "8342f66d-ff7b-4d8b-9ee0-eb2a8baaa85f",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "id": "c9c8c0ee-7fbb-4bdf-a972-cea690e17a38",
   "metadata": {},
   "source": [
    "## LP 91 HCT116 Rpb1 degron Biological Replicate 2"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "011e335d-936a-451c-93fb-b97c0a46a4ca",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d09ca4d3-7cb5-4f91-b624-5f8f9bfa052c",
   "metadata": {},
   "outputs": [],
   "source": [
    "quantiling_peaks <- function(counts_peaks, nbin = 4, condition = NULL) {\n",
    "  # Average across all tag columns\n",
    "  counts_peaks <- counts_peaks %>%\n",
    "    mutate(avg_tags = rowMeans(select(., ends_with(\"Tag\")), na.rm = TRUE))\n",
    "  \n",
    "  # If condition provided\n",
    "  if (!is.null(condition)) {\n",
    "    # Support multiple patterns combined with \"&\"\n",
    "    # Example: condition = c(\"DMSO\", \"Rpb1\") or \"DMSO&Rpb1\"\n",
    "    if (length(condition) > 1) {\n",
    "      pattern <- paste0(\"(?=.*\", condition, \")\", collapse = \"\")\n",
    "    } else if (grepl(\"&\", condition)) {\n",
    "      parts <- unlist(strsplit(condition, \"&\"))\n",
    "      pattern <- paste0(\"(?=.*\", parts, \")\", collapse = \"\")\n",
    "    } else {\n",
    "      pattern <- condition\n",
    "    }\n",
    "    \n",
    "     # Use perl = TRUE for lookaheads\n",
    "    matched_cols <- select(counts_peaks, matches(pattern, perl = TRUE))\n",
    "    \n",
    "    counts_peaks <- counts_peaks %>%\n",
    "      mutate(avg_cond = rowMeans(matched_cols, na.rm = TRUE))\n",
    "  } else {\n",
    "    counts_peaks <- counts_peaks %>%\n",
    "      mutate(avg_cond = avg_tags)\n",
    "  }\n",
    "  \n",
    "  counts_peaks <- counts_peaks %>%\n",
    "    arrange(desc(avg_cond)) %>%\n",
    "    mutate(quantile = ntile(avg_cond, nbin))\n",
    "  \n",
    "  return(counts_peaks)\n",
    "}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "3a6acecd-ab1a-4aaa-b188-9e112b94c3c9",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "id": "a31b7c9b-195b-43b2-8570-27b1e8ab2d8b",
   "metadata": {},
   "source": [
    "# HCT116 violins from counts"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "944c580d-0b46-44e5-a3c4-c39074dc5585",
   "metadata": {},
   "outputs": [
    {
     "ename": "ERROR",
     "evalue": "Error: object 'counts_H3K27ac_hg38_fly_quant' not found\n",
     "output_type": "error",
     "traceback": [
      "Error: object 'counts_H3K27ac_hg38_fly_quant' not found\nTraceback:\n",
      "1. mutate(., promoter_prox = if_else(abs(Distance.to.TSS) < 3000, \n .     \"proximal\", \"distal\"))",
      "2. mutate(., Annotation_simplified = case_when(Annotation == \"promoter-TSS\" ~ \n .     \"promoter-TSS\", Annotation == \"Intergenic\" ~ \"Intergenic\", \n .     TRUE ~ \"gene-regions\"))",
      "3. .handleSimpleError(function (cnd) \n . {\n .     watcher$capture_plot_and_output()\n .     cnd <- sanitize_call(cnd)\n .     watcher$push(cnd)\n .     switch(on_error, continue = invokeRestart(\"eval_continue\"), \n .         stop = invokeRestart(\"eval_stop\"), error = NULL)\n . }, \"object 'counts_H3K27ac_hg38_fly_quant' not found\", base::quote(eval(expr, \n .     envir)))"
     ]
    }
   ],
   "source": [
    "counts_H3K27ac_hg38_fly_quant <- counts_H3K27ac_hg38_fly_quant %>%\n",
    "  mutate(\n",
    "    Annotation_simplified = case_when(\n",
    "      Annotation == \"promoter-TSS\" ~ \"promoter-TSS\",\n",
    "      Annotation == \"Intergenic\" ~ \"Intergenic\",\n",
    "      TRUE ~ \"gene-regions\"\n",
    "    )\n",
    "  ) %>%\n",
    "mutate(promoter_prox = if_else(abs(Distance.to.TSS) < 3000, \"proximal\", \"distal\"))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d4777866-4138-4c2a-a4c4-463870df65f1",
   "metadata": {},
   "outputs": [],
   "source": [
    "options(repr.plot.width = 9, repr.plot.height = 5)\n",
    "violin_HCT116_H3K27ac_4hrTRP_distal_gene_vs_nongene <- ggplot(counts_H3K27ac_hg38_fly_quant %>% filter(quantile == 4)) + \n",
    "aes(x = interaction(Annotation_simplified,promoter_prox), \n",
    "    y = (avg_4hr_TRP/avg_0hr_TRP)) + \n",
    "geom_hline(yintercept = 1) + \n",
    "geom_violin(linewidth = 1.1, draw_quantiles = c(0.25, 0.5, 0.75)) + \n",
    "# geom_jitter(alpha = 0.3, stroke = NA) + \n",
    "scale_y_log10() + \n",
    "coord_cartesian(ylim = c(0.2, 3)) + \n",
    "theme(\n",
    "  axis.text = element_text(size = 14), \n",
    "  axis.title = element_text(size = 14), \n",
    "  legend.title = element_text(size = 14),\n",
    "  legend.text = element_text(size = 14),\n",
    "  title = element_text(size = 16),\n",
    "  strip.text = element_text(size = 14)   # facet label text size\n",
    ") + \n",
    "labs(title = \"Changes at H3K27ac Peaks in top quartile of untreated Rpb1 signal\",\n",
    "     x = \"Annotation\", y = \"H3K27ac Fold Change after 4hr TRP\") "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b1886733-9db5-4b33-a43c-26b61973cce1",
   "metadata": {},
   "outputs": [],
   "source": [
    "violin_HCT116_H3K27ac_4hrTRP_distal_gene_vs_nongene"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f89ca6e1-50be-43d1-b979-e464a147d83f",
   "metadata": {},
   "outputs": [],
   "source": [
    "counts_H3K27ac_hg38_fly_quant <- counts_H3K27ac_hg38_fly_quant %>% \n",
    "mutate(avg_4hr_IAA = (HCT116_IAA_4hr_1_H3K27ac_1.Tag + HCT116_IAA_4hr_1_H3K27ac_2.Tag + HCT116_IAA_4hr_2_H3K27ac_1.Tag + HCT116_IAA_4hr_2_H3K27ac_2.Tag)/4)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "027d5e6a-f9ff-4faa-a030-90e9696d371a",
   "metadata": {},
   "outputs": [],
   "source": [
    "options(repr.plot.width = 9, repr.plot.height = 5)\n",
    "violin_HCT116_H3K27ac_4hrIAA_distal_gene_vs_nongene <- ggplot(counts_H3K27ac_hg38_fly_quant %>% filter(quantile == 4)) + \n",
    "aes(x = interaction(Annotation_simplified, promoter_prox), \n",
    "    y = (avg_4hr_IAA/avg_0hr_TRP)) + \n",
    "geom_hline(yintercept = 1) + \n",
    "geom_violin(linewidth = 1.1, draw_quantiles = c(0.25, 0.5, 0.75)) + \n",
    "# geom_jitter(alpha = 0.3, stroke = NA) + \n",
    "scale_y_log10() + \n",
    "coord_cartesian(ylim = c(0.2, 3)) + \n",
    "theme(\n",
    "  axis.text = element_text(size = 14), \n",
    "  axis.title = element_text(size = 14), \n",
    "  legend.title = element_text(size = 14),\n",
    "  legend.text = element_text(size = 14),\n",
    "  title = element_text(size = 16),\n",
    "  strip.text = element_text(size = 14)   # facet label text size\n",
    ") + \n",
    "labs(title = \"Changes at H3K27ac Peaks in top quartile of untreated Rpb1 signal\",\n",
    "     x = \"Annotation\", y = \"H3K27ac Fold Change after 4hr IAA\") "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "51fea972-f60c-40ef-ac19-9ecdb5b15c5b",
   "metadata": {},
   "outputs": [],
   "source": [
    "violin_HCT116_H3K27ac_4hrIAA_distal_gene_vs_nongene"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b9e1ea18-c181-45fd-ac34-822d73ba4168",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "5ac467d0-6e4c-49c0-8f5f-8d67e874dbbd",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "chip-wrangler-Renv",
   "language": "R",
   "name": "chip-wrangler-renv"
  },
  "language_info": {
   "codemirror_mode": "r",
   "file_extension": ".r",
   "mimetype": "text/x-r-source",
   "name": "R",
   "pygments_lexer": "r",
   "version": "4.4.2"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
