{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "051b68d9-20fa-4a82-9157-b04451f159e4",
   "metadata": {},
   "source": [
    "# Table of Contents\n",
    "- [Processing Functions](#Process-functions)\n",
    "- [Import Data](#import-data)\n",
    "  - [Histograms](#Histograms)\n",
    "  - [Counts files](#counts)\n",
    "- [Processing](#processing)\n",
    "- [Plotting](#plotting)\n",
    "  - [HeLaS3 titration of H3K9ac](#LP78-HeLaS3-titration-of-H3K9ac)\n",
    "  - [HCT116-Rpb1 degron treatment with TRP/IAA, Biological Replicate 1](#LP-88-HCT116-Rpb1-degron-Biological-Replicate-1)\n",
    "  - [HCT116-Rpb1 degron treatment with TRP/IAA, Biological Replicate 2](#LP-91-HCT116-Rpb1-degron-Biological-Replicate-2)\n",
    "  - [HCT116-Rpb1 degron treatment with TRP/IAA, Biological Replicates 3 and 4](#LP-95-HCT116-Rpb1-degron-Biological-Replicates-3-and-4)\n",
    "- [Spike-in Normalize](#spike-in-normalize)\n",
    "  - [dual norm with confidence intervals](#dual-norm-with-confidence-intervals)\n",
    "\n",
    "---"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "5bd43ad6-d18b-465f-8ce2-1d5cc816746e",
   "metadata": {},
   "outputs": [],
   "source": [
    "library(tidyverse)\n",
    "theme_set(theme_classic())\n",
    "library(DescTools)\n",
    "library(RColorBrewer)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "9ba2e0e0-3662-4da1-a4b1-8b83a2e1083a",
   "metadata": {},
   "source": [
    "# Processing Functions"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "fa8efe20-7b8f-4825-b016-d2fa16d74823",
   "metadata": {},
   "source": [
    "## Processing HOMER Histograms\n",
    "\n",
    "The functions: `process_histograms_dm6` `process_histograms_sac3` and `process_histograms_hg38` work with histogram files generated with HOMER annotatePeaks with -hist option. More details are available [on the HOMER tutorial website](http://homer.ucsd.edu/homer/ngs/quantification.html).\n",
    "\n",
    "Making the histogram: \n",
    "\n",
    "`annotatePeaks.pl [region_to_center_on] [genome_version] -size [size_to_plot] -hist [bp_resolution] -d [tag directories] > histogram_regions_genome_datasource.txt`\n",
    "\n",
    "Example for a TSS histogram in the hg38 genome: \n",
    "\n",
    "`annotatePeaks.pl tss hg38 -size 4000 -hist 25 -d sample1-tagdir sample2-tagdir input1-tagdir input2-tagdir > hist_tss_hg38_samples12.txt`\n",
    "\n",
    "Input file: histogram_regions_genome_datasource.txt \n",
    "\n",
    "Output: dataframe in tidy (long) format, with additional columns describing the metadata originally in each sample name:\n",
    "- Cell Type\n",
    "- Treatment\n",
    "- Timepoint\n",
    "- Biological Replicate\n",
    "- Antibody (usually I have multiple IP samples from each biological sample/replicate)\n",
    "- Technical Replicate (separate library preps from the same biological sample/replicate)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "efa1dc00-a91d-4cdf-bd73-d48e34ce9a17",
   "metadata": {},
   "outputs": [],
   "source": [
    "process_histograms_dm6 <- function(x, .x) {\n",
    "    colnames(x)[1] <- \"Distance_from_tss\"\n",
    "    x <- x %>% \n",
    "      ### first two rename_with uses are specific to my sample names\n",
    "      # get rid of suffix to samples, leaving only sample name.Coverage\n",
    "    rename_with(~ gsub(\"\\\\.+\\\\.Coverage\", \".Coverage\", .x), contains(\"tagdir\")) %>% \n",
    "      # get rid of any prefix before cell type (ex: directory path)\n",
    "    rename_with(~ gsub(\".+\\\\_HCT\", \"HCT\", .x), contains(\"HCT\")) %>%\n",
    "      ### below two uses of rename_with are general to HOMER output\n",
    "    rename_with(~ gsub(\"\\\\.[[:digit:]]$\", \"_minus\", .x), contains(\"Tags\")) %>% \n",
    "    rename_with(~ gsub(\"\\\\.\\\\.\\\\.\", \"_\", .x), contains(\"Tags\"))\n",
    "    \n",
    "    # select coverage columns only\n",
    "    xcov <- x %>% select(contains(\"Coverage\"))\n",
    "    xcov$Distance_from_tss <- x$Distance_from_tss\n",
    "    \n",
    "    xcovlong <- \n",
    "    xcov %>% pivot_longer(\n",
    "      cols = -\"Distance_from_tss\", \n",
    "      names_to = \"Sample\", \n",
    "      values_to = \"Coverage\")\n",
    "    xcovlong <- xcovlong %>%\n",
    "  mutate(\n",
    "    cell = str_match(Sample, '([^_]+)(?:_[^_]+){5}$')[,2],\n",
    "    treatment = str_match(Sample, '([^_]+)(?:_[^_]+){4}$')[,2],\n",
    "    timepoint = str_remove(str_match(Sample, '([^_]+)(?:_[^_]+){3}$')[,2], \"TSA\"),\n",
    "    biorep = str_match(Sample, '([^_]+)(?:_[^_]+){2}$')[,2],\n",
    "    antibody = str_match(Sample, '([^_]+)(?:_[^_]+){1}$')[,2],\n",
    "    techrep = str_match(Sample, '([^_]+)(?:_[^_]+){0}$')[,2]\n",
    "  )\n",
    "  }\n",
    "\n",
    "process_histograms_sac3 <- function(x, .x) {\n",
    "    colnames(x)[1] <- \"Distance_from_tss\"\n",
    "    x <- x %>% \n",
    "      ### first two rename_with uses are specific to my sample names\n",
    "      # get rid of suffix to samples, leaving only sample name.Coverage\n",
    "    rename_with(~ gsub(\".concat.+Coverage\", \".Coverage\", .x), contains(\"tagdir\")) %>% \n",
    "  #  rename_with(~ gsub(\"_H3K9ac\", \"_1_H3K9ac\", .x), contains(\"Hela\")) %>%\n",
    " #   rename_with(~ gsub(\"_input\", \"_1_input\", .x), contains(\"sync\")) %>%\n",
    "      # get rid of any prefix before cell type (ex: directory path)\n",
    "    rename_with(~ gsub(\".+\\\\_HCT\", \"HCT\", .x), contains(\"HCT\")) %>%\n",
    "    # get rid of any suffix (ex: S2, S4)\n",
    "      ### below two uses of rename_with are general to HOMER output\n",
    "    rename_with(~ gsub(\"\\\\.[[:digit:]]$\", \"_minus\", .x), contains(\"Tags\")) %>% \n",
    "    rename_with(~ gsub(\"\\\\.\\\\.\\\\.\", \"_\", .x), contains(\"Tags\"))\n",
    "    \n",
    "    # select coverage columns only\n",
    "    xcov <- x %>% select(contains(\"Coverage\"))\n",
    "    xcov$Distance_from_tss <- x$Distance_from_tss\n",
    "    \n",
    "    xcovlong <- \n",
    "    xcov %>% pivot_longer(\n",
    "      cols = -\"Distance_from_tss\", \n",
    "      names_to = \"Sample\", \n",
    "      values_to = \"Coverage\")\n",
    "    xcovlong <- xcovlong %>%\n",
    "  mutate(\n",
    "    cell = str_match(Sample, '([^_]+)(?:_[^_]+){5}$')[,2],\n",
    "    treatment = str_match(Sample, '([^_]+)(?:_[^_]+){4}$')[,2],\n",
    "    timepoint = str_remove(str_match(Sample, '([^_]+)(?:_[^_]+){3}$')[,2], \"TSA\"),\n",
    "    biorep = str_match(Sample, '([^_]+)(?:_[^_]+){2}$')[,2],\n",
    "    antibody = str_match(Sample, '([^_]+)(?:_[^_]+){1}$')[,2],\n",
    "    techrep = str_match(Sample, '([^_]+)(?:_[^_]+){0}$')[,2]\n",
    "  )\n",
    "  }\n",
    "\n",
    "process_histograms_hg38 <- function(x, .x) {\n",
    "    colnames(x)[1] <- \"Distance_from_tss\"\n",
    "    x <- x %>% \n",
    "      ### first two rename_with uses are specific to my sample names\n",
    "      # get rid of suffix to samples, leaving only sample name.Coverage\n",
    "    rename_with(~ gsub(\".concat.hg38.tagdir.Coverage\", \".Coverage\", .x), contains(\"HCT\")) %>% \n",
    "      # get rid of any prefix before cell type (ex: directory path)\n",
    "    rename_with(~ gsub(\".+\\\\_HCT\", \"HCT\", .x), contains(\"HCT\")) %>% \n",
    "      ### below two uses of rename_with are general to HOMER output\n",
    "    rename_with(~ gsub(\"\\\\.[[:digit:]]$\", \"_minus\", .x), contains(\"Tags\")) %>% \n",
    "    rename_with(~ gsub(\"\\\\.\\\\.\\\\.\", \"_\", .x), contains(\"Tags\"))\n",
    "    \n",
    "    # select coverage columns only\n",
    "    xcov <- x %>% select(contains(\"Coverage\"))\n",
    "    xcov$Distance_from_tss <- x$Distance_from_tss\n",
    "    \n",
    "    xcovlong <- \n",
    "    xcov %>% pivot_longer(\n",
    "      cols = -\"Distance_from_tss\", \n",
    "      names_to = \"Sample\", \n",
    "      values_to = \"Coverage\")\n",
    "    \n",
    "    xcovlong <- xcovlong %>%\n",
    "  mutate(\n",
    "    cell = str_match(Sample, '([^_]+)(?:_[^_]+){5}$')[,2],\n",
    "    treatment = str_match(Sample, '([^_]+)(?:_[^_]+){4}$')[,2],\n",
    "    timepoint = str_remove(str_match(Sample, '([^_]+)(?:_[^_]+){3}$')[,2], \"TSA\"),\n",
    "    biorep = str_match(Sample, '([^_]+)(?:_[^_]+){2}$')[,2],\n",
    "    antibody = str_match(Sample, '([^_]+)(?:_[^_]+){1}$')[,2],\n",
    "    techrep = str_match(Sample, '([^_]+)(?:_[^_]+){0}$')[,2]\n",
    "  )\n",
    "  }"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "ad3daec4-afb5-419b-9e8f-24c3c9650a0b",
   "metadata": {},
   "source": [
    "## Processing HOMER counts files\n",
    "\n",
    "The function `process_counts_annotpeaks()` takes HOMER annotatePeaks files generated without the -hist option. Instead of generating average Coverage across all regions specified, the resulting annotated counts file is a matrix, with rows for each region. The first 19 columns are metadata describing each region, the remaining columns are the Read Normalized Tag Counts for each sample at each region.\n",
    "\n",
    "Relevant Metadata Columns: \n",
    "1. **PeakID**\n",
    "2. **Chr**\n",
    "3. **Start**\n",
    "4. **End**\n",
    "5. Strand - strand that nearest annotated gene/TSS is located on, not necessarily strand of your data! Remember ChIP-seq/ATAC-seq is unstranded by definition. \n",
    "6. Not Used - ignore\n",
    "7. Focus.Ratio.Region.Size\n",
    "8. **Annotation** - if annotating on TSS regions, all will be promoter-TSS\n",
    "9. Detailed Annotation\n",
    "10. **Distance to TSS** - if annotating on TSS regions, all will be 0."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "76f42e6a-3082-40e0-8897-aa11c6bf68c4",
   "metadata": {},
   "outputs": [],
   "source": [
    "process_counts_annotpeaks <- function(counts_annotpeaks, .x) {\n",
    "  colnames(counts_annotpeaks)[1] <- \"PeakID\"\n",
    "  counts_annotpeaks <- counts_annotpeaks %>% \n",
    "    rename_with(~ gsub(\".hg38.tagdir\", \"\", .x), contains(\"tagdir\")) %>% \n",
    "    rename_with(~ gsub(\".dedup.+.Tag\", \".Tag\", .x), contains(\"tagdir\")) %>%\n",
    "    rename_with(~ gsub(\"\\\\.concat\", \"\", .x), contains(\"Tag\")) %>% \n",
    "    rename_with(~ gsub(\".+HCT\", \"HCT\", .x), contains(\"Tag\")) %>%\n",
    "    rename_with(~ gsub(\".+Hela\", \"Hela\", .x), contains(\"Tag\")) %>%\n",
    "    rename_with(~ gsub(\"\\\\.[[:digit:]]$\", \"_minus\", .x), contains(\"Tag\")) %>% \n",
    "    rename_with(~ gsub(\"\\\\.Count.+\", \"\", .x), contains(\"Tag\"))\n",
    "} "
   ]
  },
  {
   "cell_type": "markdown",
   "id": "7be74e72-7c2e-4f40-8966-685795c686c9",
   "metadata": {},
   "source": [
    "# import data"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "e777fa75-3dd9-487d-9141-3ac156a170c2",
   "metadata": {},
   "source": [
    "## Histograms"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "73a44c2b-6a27-4cee-ad1b-f2b63dddc929",
   "metadata": {},
   "source": [
    "There are 3 types of histogram files: \n",
    "1. TSS histograms: useful for H3K27ac, H3K4me3, Rpb1, other active histone marks...\n",
    "2. Peak histograms: calculated around HOMER-defined peak sets, can do for individual IP targets\n",
    "3. RNA histograms: calculated around gene bodies, useful for H3K36me3, Rpb1\n",
    "\n",
    "Note that gene bodies can also be grouped by activity (calculated here by untreated/control Rpb1 levels) or gene length"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "669a6045-9c1b-4acf-a4e3-31ca22263c76",
   "metadata": {},
   "outputs": [],
   "source": [
    "# LP 78 - Mitotic H3K9ac titration\n",
    "hist_tss_dm6_LP78 <- read.delim(\"LP_78/data/dm6_data/dm6_nodups_tagdirs/hist_tss_dm6_LP78.txt\")\n",
    "hist_tss_sac3_LP78 <- read.delim(\"LP_78/data/sac3_data/sac3_nodups_tagdirs/hist_tss_sac3_LP78.txt\")\n",
    "hist_tss_hg38_LP78 <- read.delim(\"LP_78/data/hg38_data/hg38_nodups_tagdirs/hist_tss_hg38_LP78_cleaned.txt\")\n",
    "hist_peaks_hg38_LP78 <- read.delim(\"LP_78/data/hg38_data/hg38_nodups_tagdirs/hist_mergedpeaks_hg38_LP78.txt\")\n",
    "\n",
    "# LP 88 - HCT116-Rpb1 degron biological replicate 1\n",
    "hist_rna_dm6_LP88 <- read.delim(\"LP_88_combined_fastqs/dm6_data/dm6_tagdirs/hist_rna_dm6_LP88.txt\")\n",
    "hist_tss_dm6_LP88 <- read.delim(\"LP_88_combined_fastqs/dm6_data/dm6_tagdirs/hist_tss_dm6_LP88.txt\")\n",
    "hist_K4me1_dm6_LP88 <- read.delim(\"LP_88_combined_fastqs/dm6_data/dm6_tagdirs/hist_H3K4me1peaks_dm6_LP88.txt\")\n",
    "hist_K36me3_dm6_LP88 <- read.delim(\"LP_88_combined_fastqs/dm6_data/dm6_tagdirs/hist_H3K36me3peaks_dm6_LP88.txt\")\n",
    "hist_rna_sac3_LP88 <- read.delim(\"LP_88_combined_fastqs/sac3_data/sac3_tagdirs/hist_rna_sac3_LP88.txt\")\n",
    "hist_tss_sac3_LP88 <- read.delim(\"LP_88_combined_fastqs/sac3_data/sac3_tagdirs/hist_tss_sac3_LP88.txt\")\n",
    "hist_tss_hg38_LP88 <- read.delim(\"LP_88_combined_fastqs/hg38_data/hg38_tagdirs/hist_tss_hg38_LP88.txt\")\n",
    "hist_K27ac_hg38_LP88 <- read.delim(\"LP_88_combined_fastqs/hg38_data/hg38_tagdirs/hist_H3K27acpeaks_hg38_LP88.txt\")\n",
    "\n",
    "# LP 91 - HCT116-Rpb1 degron biological replicate 2, technical replicate 1/2\n",
    "hist_rna_dm6_LP91 <- read.delim(\"LP_91/dm6_data/dm6_tagdirs/hist_rna_dm6_LP91.txt\")\n",
    "hist_tss_dm6_LP91 <- read.delim(\"LP_91/dm6_data/dm6_tagdirs/hist_tss_dm6_LP91.txt\")\n",
    "hist_K4me1_dm6_LP91 <- read.delim(\"LP_91/dm6_data/dm6_tagdirs/hist_H3K4me1peaks_dm6_LP91.txt\")\n",
    "hist_K36me3_dm6_LP91 <- read.delim(\"LP_91/dm6_data/dm6_tagdirs/hist_H3K36me3peaks_dm6_LP91.txt\")\n",
    "hist_rna_sac3_LP91 <- read.delim(\"LP_91/sac3_data/sac3_tagdirs/hist_rna_sac3_LP91.txt\")\n",
    "hist_tss_sac3_LP91 <- read.delim(\"LP_91/sac3_data/sac3_tagdirs/hist_tss_sac3_LP91.txt\")\n",
    "\n",
    "hist_shortrna_hg38_LP91 <- read.delim(\"LP_91/hg38_data/hg38_tagdirs/hist_5to20kb_hg38_LP91.txt\")\n",
    "hist_tss_hg38_LP91 <- read.delim(\"LP_91/hg38_data/hg38_tagdirs/hist_tss_hg38_LP91.txt\")\n",
    "hist_rna_hg38_LP91 <- read.delim(\"LP_91/hg38_data/hg38_tagdirs/hist_rna_hg38_LP91.txt\")\n",
    "\n",
    "# LP 95 - HCT116-Rpb1 degron biological replicates 3 and 4\n",
    "hist_rna_dm6_LP95 <- read.delim(\"~/data/250627_nextseq2000_LP95/dm6_data/dm6_tagdirs/hist_rna_dm6_LP95.txt\")\n",
    "hist_tss_dm6_LP95 <- read.delim(\"~/data/250627_nextseq2000_LP95/dm6_data/dm6_tagdirs/hist_tss_dm6_LP95.txt\")\n",
    "hist_rna_sac3_LP95 <- read.delim(\"~/data/250627_nextseq2000_LP95/sac3_data/sac3_tagdirs/hist_rna_sac3_LP95.txt\")\n",
    "hist_tss_sac3_LP95 <- read.delim(\"~/data/250627_nextseq2000_LP95/sac3_data/sac3_tagdirs/hist_tss_sac3_LP95.txt\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "a1fb8630-1a74-4c45-8f70-e87d412c0eda",
   "metadata": {},
   "source": [
    "## Counts"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "826535b5-e04b-49f3-a7d4-2fb774a843c9",
   "metadata": {},
   "outputs": [],
   "source": [
    "counts_chr19_hg38_LP78 <- read.delim(\"~/data/experiments/LP_78/data/hg38_data/counts_chr19_1kb_hg38_LP78_readnormvsspike.txt\")\n",
    "\n",
    "counts_tss_hg38_LP91 <- read.delim(\"LP_91/hg38_data/hg38_tagdirs/counts_tss_hg38_LP88_LP91.txt\")\n",
    "counts_tss_hg38_LP95 <- read.delim(\"~/data/250627_nextseq2000_LP95/hg38_data/hg38_tagdirs/counts_tss_hg38_LP95.txt\")\n",
    "counts_allHCT_4hrTRP_H3K27ac <- read.delim(\"~/data/dualspike_natgen_resubmit1/counts_allHCT_bioreps_4hrTRP_DMSO_H3K27ac_readnorm_alldata.txt\")\n",
    "\n",
    "counts_rna_hg38_spikenorm <- read.delim(\"~/data/dualspike_natgen_resubmit1/counts_rna_hg38_Rpb1_H3K36me3_spikenorm.txt\")\n",
    "\n",
    "counts_Hela_tss <- read.delim(\"~/data/experiments/LP_81_combined_fastqs/hg38_data/hg38_normalized_tagdirs/counts_alltss_hg38_Hela_spikenorm.txt\")\n",
    "counts_Hela <- read.delim(\"~/data/dualspike_natgen_resubmit1/counts_allHela_H3K27ac_hg38_LP100_LP81_spikenorm.txt\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "b7790bb3-d6c8-4f95-902e-67db91a1e93b",
   "metadata": {},
   "source": [
    "# Processing"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "5cf96fb4-967f-47a0-947f-20731a64957b",
   "metadata": {},
   "outputs": [],
   "source": [
    "hist_tss_dm6_LP78_long <- process_histograms_dm6(hist_tss_dm6_LP78)\n",
    "hist_tss_sac3_LP78_long <- process_histograms_sac3(hist_tss_sac3_LP78)\n",
    "hist_tss_hg38_LP78_long <- process_histograms_hg38(hist_tss_hg38_LP78)\n",
    "hist_peaks_hg38_LP78_long <- process_histograms_hg38(hist_peaks_hg38_LP78)\n",
    "\n",
    "hist_tss_dm6_LP81_long <- process_histograms_dm6(hist_tss_dm6_LP81)\n",
    "hist_tss_sac3_LP81_long <- process_histograms_dm6(hist_tss_sac3_LP81)\n",
    "hist_tss_hg38_LP81_long <- process_histograms_dm6(hist_tss_hg38_LP81)\n",
    "hist_K4me1_dm6_LP81_long <- process_histograms_dm6(hist_K4me1_dm6_LP81)\n",
    "\n",
    "hist_rna_dm6_LP88_long <- process_histograms_dm6(hist_rna_dm6_LP88)\n",
    "hist_tss_dm6_LP88_long <- process_histograms_dm6(hist_tss_dm6_LP88)\n",
    "hist_K36me3_dm6_LP88_long <- process_histograms_dm6(hist_K36me3_dm6_LP88)\n",
    "hist_K4me1_dm6_LP88_long <- process_histograms_dm6(hist_K4me1_dm6_LP88)\n",
    "hist_rna_sac3_LP88_long <- process_histograms_dm6(hist_rna_sac3_LP88)\n",
    "hist_tss_sac3_LP88_long <- process_histograms_dm6(hist_tss_sac3_LP88)\n",
    "hist_tss_hg38_LP88_long <- process_histograms_dm6(hist_tss_hg38_LP88)\n",
    "hist_over100kb_hg38_LP88_long <- process_histograms_hg38(hist_over100kb_hg38_LP88)\n",
    "\n",
    "hist_rna_dm6_LP91_long <- process_histograms_dm6(hist_rna_dm6_LP91)\n",
    "hist_tss_dm6_LP91_long <- process_histograms_dm6(hist_tss_dm6_LP91)\n",
    "hist_K36me3_dm6_LP91_long <- process_histograms_dm6(hist_K36me3_dm6_LP91)\n",
    "hist_K4me1_dm6_LP91_long <- process_histograms_dm6(hist_K4me1_dm6_LP91)\n",
    "hist_rna_sac3_LP91_long <- process_histograms_dm6(hist_rna_sac3_LP91)\n",
    "hist_tss_sac3_LP91_long <- process_histograms_dm6(hist_tss_sac3_LP91)\n",
    "\n",
    "hist_shortrna_hg38_LP91_long <- process_histograms_hg38(hist_shortrna_hg38_LP91)\n",
    "hist_tss_hg38_LP91_long <- process_histograms_hg38(hist_tss_hg38_LP91)\n",
    "\n",
    "hist_tss_dm6_LP92_long <- process_histograms_dm6(hist_tss_dm6_LP92)\n",
    "hist_tss_sac3_LP92_long <- process_histograms_dm6(hist_tss_sac3_LP92)\n",
    "\n",
    "hist_rna_dm6_LP95_long <- process_histograms_dm6(hist_rna_dm6_LP95)\n",
    "hist_tss_dm6_LP95_long <- process_histograms_dm6(hist_tss_dm6_LP95)\n",
    "hist_rna_sac3_LP95_long <- process_histograms_dm6(hist_rna_sac3_LP95)\n",
    "hist_tss_sac3_LP95_long <- process_histograms_dm6(hist_tss_sac3_LP95)\n",
    "\n",
    "hist_shortrna_hg38_LP95_long <- process_histograms_hg38(hist_shortrna_hg38_LP95)\n",
    "hist_over100kb_hg38_LP95_long <- process_histograms_hg38(hist_over100kb_hg38_LP95)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "9fa23cf8-5280-4014-941d-b9a04c672d75",
   "metadata": {},
   "outputs": [],
   "source": [
    "counts_chr19_hg38_LP78 <- process_counts_annotpeaks(counts_chr19_hg38_LP78)\n",
    "counts_tss_hg38_LP91 <- process_counts_annotpeaks(counts_tss_hg38_LP91)\n",
    "counts_tss_hg38_LP95 <- process_counts_annotpeaks(counts_tss_hg38_LP95)\n",
    "\n",
    "counts_allHCT_4hrTRP_H3K27ac <- process_counts_annotpeaks(counts_allHCT_4hrTRP_H3K27ac)\n",
    "\n",
    "counts_Hela <- process_counts_annotpeaks(counts_Hela)\n",
    "\n",
    "counts_rna_hg38_spikenorm <- process_counts_annotpeaks(counts_rna_hg38_spikenorm)\n",
    "\n",
    "counts_Hela_tss <- process_counts_annotpeaks(counts_Hela_tss)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "8ee882ed-dc81-43e0-afac-cef80cef84e5",
   "metadata": {},
   "outputs": [],
   "source": [
    "counts_rna_hg38_spikenorm$Annotation <- gsub(\"\\\\(.*\", \"\", as.character(counts_rna_hg38_spikenorm$Annotation))\n",
    "counts_rna_hg38_spikenorm$Annotation <- gsub(\" \", \"\", as.character(counts_rna_hg38_spikenorm$Annotation))"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "722f1708-c123-41e4-ae8e-e3df7cd0ca03",
   "metadata": {},
   "source": [
    "# Figure 1"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "6323e09f-aeb4-4f93-9cb4-1a449215c6df",
   "metadata": {},
   "outputs": [],
   "source": [
    "LP78_hist_tss_dm6 <- ggplot(hist_tss_dm6_LP78_long %>% \n",
    "                            arrange((as.factor(as.numeric(str_remove(timepoint, \"inter\"))))) %>% \n",
    "                            filter(grepl(\"Hela\", Sample))) + \n",
    "aes(x = Distance_from_tss, y = Coverage, \n",
    "    group= Sample, color = as.factor(as.numeric(str_remove(timepoint, \"inter\")))) +\n",
    "  scale_color_manual(\n",
    "      values = colorRampPalette(brewer.pal(9, \"Greys\"))(7)[2:7],\n",
    "      name = \"% Human Interphase Cells\") +\n",
    "geom_line(linewidth = 1.1, alpha = 0.8) + \n",
    "facet_wrap(~ antibody) + \n",
    "labs(title = \"Fly signal at TSSs\", x = \"Distance from TSS Center\")\n",
    "\n",
    "# Open SVG device (base R)\n",
    "#svg(\"LP78_hist_tss_dm6.svg\", width = 16, height = 7)\n",
    "\n",
    "# Print the ggplot\n",
    "print(LP78_hist_tss_dm6) \n",
    "\n",
    "# Close the device to save the file\n",
    "dev.off()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "ffaae743-87ea-4e2b-9bbe-da70d06fcc56",
   "metadata": {},
   "outputs": [],
   "source": [
    "hist_tss_sac3_LP78_long %>% arrange((as.factor(as.numeric(str_remove(timepoint, \"inter\"))))) %>% head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "401588e6-dc95-4ffb-b681-65877462a236",
   "metadata": {},
   "outputs": [],
   "source": [
    "options(repr.plot.width = 16, repr.plot.height = 7)\n",
    "LP78_hist_tss_sac3 <- ggplot(hist_tss_sac3_LP78_long %>% \n",
    "       arrange((as.factor(as.numeric(str_remove(timepoint, \"inter\"))))) %>% \n",
    "       filter(grepl(\"Hela\", Sample))) + \n",
    "aes(x = Distance_from_tss, y = Coverage, \n",
    "    group= Sample, color = as.factor(as.numeric(str_remove(timepoint, \"inter\")))) +\n",
    "  scale_color_manual(\n",
    "      values = colorRampPalette(brewer.pal(9, \"Reds\"))(7)[2:7],\n",
    "      name = \"% Human Interphase Cells\") +\n",
    "geom_line(linewidth = 1.1, alpha = 0.6) + \n",
    "facet_wrap(~ antibody, scales = \"free\") + \n",
    "labs(title = \"Yeast signal at TSSs\", x = \"Distance from TSS Center\")\n",
    "\n",
    "# Open SVG device (base R)\n",
    "#svg(\"LP78_hist_tss_sac3.svg\", width = 16, height = 7)\n",
    "\n",
    "# Print the ggplot\n",
    "print(LP78_hist_tss_sac3) \n",
    "\n",
    "# Close the device to save the file\n",
    "dev.off()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b02acaf2-3e54-45b4-90cc-5afe49815ea3",
   "metadata": {},
   "outputs": [],
   "source": [
    "ggplot(hist_tss_hg38_LP78_long %>% \n",
    "                            arrange((as.factor(as.numeric(str_remove(timepoint, \"inter\"))))) %>% \n",
    "                            filter(grepl(\"Hela\", Sample))) + \n",
    "aes(x = Distance_from_tss, y = Coverage, \n",
    "    group= Sample, color = as.factor(as.numeric(str_remove(timepoint, \"inter\")))) +\n",
    "  scale_color_manual(\n",
    "      values = colorRampPalette(brewer.pal(9, \"Blues\"))(7)[2:7],\n",
    "      name = \"% Human Interphase Cells\") +\n",
    "geom_line(linewidth = 1.1, alpha = 0.8) + \n",
    "facet_wrap(~ antibody) + \n",
    "labs(title = \"Human signal at TSSs\", x = \"Distance from TSS Center\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "63594476-3f17-4a80-a355-5bde859944af",
   "metadata": {},
   "source": [
    "### LP78 H3K9ac titration scatterplots"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "ccf474dc-6f02-4382-8e78-ec96e3eaeb03",
   "metadata": {},
   "outputs": [],
   "source": [
    "counts_chr19_hg38_LP78_avg <- counts_chr19_hg38_LP78 %>%\n",
    "  # Pivot longer: collect all tag columns\n",
    "  pivot_longer(\n",
    "    cols = ends_with(\".Tag\"),\n",
    "    names_to = \"SampleColumn\",\n",
    "    values_to = \"TagValue\"\n",
    "  ) %>%\n",
    "  \n",
    "  # Extract normalization type\n",
    "  mutate(\n",
    "    Normalization = case_when(\n",
    "      str_detect(SampleColumn, \"\\\\.normalized\\\\.tagdir\\\\.Tag$\") ~ \"spike_in\",\n",
    "      str_detect(SampleColumn, \"\\\\.hg38\\\\.GCcheck\\\\.tagdir\\\\.Tag$\") ~ \"read_depth\",\n",
    "      TRUE ~ NA_character_\n",
    "    ),\n",
    "    # remove normalization suffix from name\n",
    "    Sample = SampleColumn %>%\n",
    "      str_remove(\"\\\\.normalized\\\\.tagdir\\\\.Tag$\") %>%\n",
    "      str_remove(\"\\\\.hg38\\\\.GCcheck\\\\.tagdir\\\\.Tag$\")\n",
    "  ) %>%\n",
    "  select(-SampleColumn) %>%\n",
    "  \n",
    "  # Extract replicate number (assuming _1/_2/_3)\n",
    "  mutate(\n",
    "    Replicate = str_extract(Sample, \"_\\\\d+$\"),\n",
    "    SampleGroup = str_remove(Sample, \"_\\\\d+$\")\n",
    "  ) %>%\n",
    "  \n",
    "  # Average technical replicates per Peak × SampleGroup × Normalization\n",
    "  group_by(across(c(PeakID, Chr, Start, End, Strand, Peak.Score,\n",
    "                    Focus.Ratio.Region.Size, Annotation, Detailed.Annotation,\n",
    "                    Distance.to.TSS, Nearest.PromoterID, Entrez.ID,\n",
    "                    Nearest.Unigene, Nearest.Refseq, Nearest.Ensembl,\n",
    "                    Gene.Name, Gene.Alias, Gene.Description, Gene.Type,\n",
    "                    SampleGroup, Normalization))) %>%\n",
    "  summarise(MeanTag = mean(TagValue, na.rm = TRUE), .groups = \"drop\") %>%\n",
    "  \n",
    "  # Pivot wider: one column per SampleGroup × Normalization\n",
    "  pivot_wider(\n",
    "    names_from = c(SampleGroup, Normalization),\n",
    "    names_glue = \"{SampleGroup}.{Normalization}\",\n",
    "    values_from = MeanTag\n",
    "  )"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "6ee25eb3-a2c1-475a-b2dd-4bd5fe794ddd",
   "metadata": {},
   "outputs": [],
   "source": [
    "options(repr.plot.width = 5.7, repr.plot.height = 5)\n",
    "\n",
    "ggplot(counts_chr19_hg38_LP78_avg) + \n",
    "geom_abline() +\n",
    "geom_point(aes(x = HelaS3_100sync_0inter_1_H3K9ac.read_depth, \n",
    "    y = HelaS3_0sync_100inter_1_H3K9ac.read_depth), alpha = 0.05, color = \"grey\") + \n",
    "geom_point(aes(x = HelaS3_100sync_0inter_1_H3K9ac.spike_in, \n",
    "    y = HelaS3_0sync_100inter_1_H3K9ac.spike_in), alpha = 0.05, color = \"black\") + \n",
    "#geom_smooth(aes(x = HelaS3_100sync_0inter_1_H3K9ac.spike_in,  y = HelaS3_25sync_75inter_1_H3K9ac.spike_in), \n",
    "#    method = \"lm\", se = FALSE, color = \"#93c68c\", size = 1.2) +\n",
    "scale_x_log10() + \n",
    "scale_y_log10() + \n",
    "coord_cartesian(xlim = c(0.2, 650), ylim = c(0.2, 650)) +\n",
    "theme(axis.text = element_text(size=20), \n",
    "        axis.title=element_text(size=20)) + \n",
    "labs(x = \"Avg 0% Interphase Signal\", y = \"Avg 100% Signal\") "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d08d0ae4-9890-4d97-9872-df81259a7f32",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "id": "589bdd7a-a015-49de-bac5-bb2e5f14ed3a",
   "metadata": {},
   "source": [
    "# Figure 3"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "eb44b000-2fa4-464d-b94d-ac0dadd6db29",
   "metadata": {},
   "source": [
    "### *H. sapiens* data\n",
    "\n",
    "We quantify signal at regions biologically relevant for each antibody:\n",
    "- H3K27ac: at TSSs or H3K27ac peaks\n",
    "- H3K4me3: at TSSs or H3K4me3 peaks\n",
    "- H3K4me1: at H3K4me1 peaks\n",
    "- H3K36me3: at gene bodies or H3K36me3 peaks\n",
    "- Rpb1: at TSSs or Rpb1 peaks"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f1baa090-2b51-4689-a276-9289091bf47a",
   "metadata": {},
   "outputs": [],
   "source": [
    "options(repr.plot.width = 14, repr.plot.height = 9)\n",
    "ggplot(hist_tss_hg38_LP88_long) + \n",
    "aes(x = Distance_from_tss, y = Coverage, \n",
    "    group= Sample, color = timepoint) +\n",
    "  scale_color_manual(\n",
    "    values = c(\"firebrick\",\"slateblue\",\"goldenrod\"),\n",
    "    name = \"Timepoint\") + \n",
    "geom_line(linewidth = 1.1, alpha = 0.8) + \n",
    "facet_wrap(~ antibody, scales = \"free\") + \n",
    "labs(title = \"Human signal at TSS\", x = \"Distance from TSS Center\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b78a6220-2dff-403f-bbec-0e9582c0738c",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "id": "3be4916a-6bbe-4840-a01d-ba93fb0eeffd",
   "metadata": {},
   "source": [
    "# Calculate spike-in SNRs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "7b861eca-8a6c-47ed-bee4-a1a52abb7abb",
   "metadata": {},
   "outputs": [],
   "source": [
    "histograms_signalarea_calculate_sizegiven <- function(input_df, Sample, Coverage, colname1) {\n",
    "  # ensym() converts strings into symbols, so function inputs can be understood\n",
    "  ### within ggplot\n",
    "  Sample <- rlang::ensym(Sample)\n",
    "  Coverage <- rlang::ensym(Coverage)\n",
    "  colname1 <- rlang::ensym(colname1)\n",
    "  # get a list of unique sample names\n",
    "  samples <- unique(input_df$Sample)\n",
    "\n",
    "  # make sure first column is named properly\n",
    "  colnames(input_df)[1] <- \"Distance_from_center\"  \n",
    "    \n",
    "  # filter to only calculate between -1000 and 1000\n",
    "  input_df <- input_df %>%\n",
    "    dplyr::filter(Distance_from_center >= 0.04 & Distance_from_center <= 1)\n",
    "    \n",
    "  # get x variable\n",
    "  x <- unique(input_df[[1]])\n",
    "    \n",
    "  # initialize empty matrix, then use to make dataframe output\n",
    "  AUC_peaks <- matrix(data = \"\", nrow = length(samples), ncol = 1)\n",
    "  AUC_peaks_df <- data.frame(AUC_peaks, row.names = samples)\n",
    "\n",
    "  # fill in AUC dataframe for each sample\n",
    "  for (i in 1:length(samples)) {\n",
    "\n",
    "    y <- input_df |>\n",
    "      dplyr::filter(Distance_from_center >= 0.04 & Distance_from_center <= 1) |>\n",
    "      dplyr::filter(Sample == samples[i]) |>\n",
    "      dplyr::select(Coverage)\n",
    "    y <- dplyr::pull(y, Coverage)\n",
    "\n",
    "    AUC_peaks_df[i, ] <- DescTools::AUC(x, y, method = c(\"trapezoid\"))\n",
    "\n",
    "  }\n",
    "  # get data in numeric form\n",
    "  AUC_peaks_df <- as.data.frame(sapply(AUC_peaks_df, as.numeric))\n",
    "\n",
    "  # add in the treatment column for plotting\n",
    "  treatment_df <- input_df |>\n",
    "    dplyr::filter(Distance_from_center == \"0.04\") |>\n",
    "   dplyr::select(treatment)\n",
    "\n",
    "  treatment_vect <- dplyr::pull(treatment_df, treatment)\n",
    "\n",
    "  # graph labels in title custom\n",
    "  cell_type <- unique(input_df[[3]])\n",
    "  AUC_name <- colnames(AUC_peaks_df[1])\n",
    "    \n",
    "    AUC_peaks_df$ID <- samples\n",
    "   \n",
    "    AUC_peaks_df\n",
    "}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d172b07d-8af3-4d5d-b271-fd989b186287",
   "metadata": {},
   "outputs": [],
   "source": [
    "histograms_signalarea_calculatetss <- function(input_df, Sample, Coverage, colname1) {\n",
    "  # ensym() converts strings into symbols, so function inputs can be understood\n",
    "  ### within ggplot\n",
    "  Sample <- rlang::ensym(Sample)\n",
    "  Coverage <- rlang::ensym(Coverage)\n",
    "  colname1 <- rlang::ensym(colname1)\n",
    "  # get a list of unique sample names\n",
    "  samples <- unique(input_df$Sample)\n",
    "\n",
    "  # make sure first column is named properly\n",
    "  colnames(input_df)[1] <- \"Distance_from_center\"  \n",
    "    \n",
    "  # filter to only calculate between -1000 and 1000\n",
    "  input_df <- input_df %>%\n",
    "    dplyr::filter(Distance_from_center >= -200 & Distance_from_center <= 700)\n",
    "    \n",
    "  # get x variable\n",
    "  x <- unique(input_df[[1]])\n",
    "    \n",
    "  # initialize empty matrix, then use to make dataframe output\n",
    "  AUC_peaks <- matrix(data = \"\", nrow = length(samples), ncol = 1)\n",
    "  AUC_peaks_df <- data.frame(AUC_peaks, row.names = samples)\n",
    "\n",
    "  # fill in AUC dataframe for each sample\n",
    "  for (i in 1:length(samples)) {\n",
    "\n",
    "    y <- input_df |>\n",
    "      dplyr::filter(Distance_from_center >= -200 & Distance_from_center <= 700) |>\n",
    "      dplyr::filter(Sample == samples[i]) |>\n",
    "      dplyr::select(Coverage)\n",
    "    y <- dplyr::pull(y, Coverage)\n",
    "\n",
    "    AUC_peaks_df[i, ] <- DescTools::AUC(x, y, method = c(\"trapezoid\"))\n",
    "\n",
    "  }\n",
    "  # get data in numeric form\n",
    "  AUC_peaks_df <- as.data.frame(sapply(AUC_peaks_df, as.numeric))\n",
    "\n",
    "  # add in the treatment column for plotting\n",
    "  treatment_df <- input_df |>\n",
    "    dplyr::filter(Distance_from_center == \"-200\") |>\n",
    "   dplyr::select(treatment)\n",
    "\n",
    "  treatment_vect <- dplyr::pull(treatment_df, treatment)\n",
    "\n",
    "  # graph labels in title custom\n",
    "  cell_type <- unique(input_df[[3]])\n",
    "  AUC_name <- colnames(AUC_peaks_df[1])\n",
    "    \n",
    "    AUC_peaks_df$ID <- samples\n",
    "   \n",
    "    AUC_peaks_df\n",
    "}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d41c764f-c0ad-4e1b-b141-569f63c46643",
   "metadata": {},
   "outputs": [],
   "source": [
    "fly_tss_signal_LP78 <- histograms_signalarea_calculatetss(hist_tss_dm6_LP78_long, colname1 = colname1)\n",
    "yeast_tss_signal_LP78 <- histograms_signalarea_calculatetss(hist_tss_sac3_LP78_long, colname1 = colname1)\n",
    "\n",
    "fly_tss_signal_LP81 <- histograms_signalarea_calculatetss(hist_tss_dm6_LP81_long, colname1 = colname1)\n",
    "yeast_tss_signal_LP81 <- histograms_signalarea_calculatetss(hist_tss_sac3_LP81_long, colname1 = colname1)\n",
    "fly_K4me1_signal_LP81 <- histograms_signalarea_calculatetss(hist_K4me1_dm6_LP81_long, colname1 = colname1)\n",
    "\n",
    "fly_rna_signal_LP88 <- histograms_signalarea_calculate_sizegiven(hist_rna_dm6_LP88_long, colname1 = colname1)\n",
    "fly_K36me3_signal_LP88 <- histograms_signalarea_calculate_sizegiven(hist_K36me3_dm6_LP88_long, colname1 = colname1)\n",
    "fly_K4me1_signal_LP88 <- histograms_signalarea_calculate_sizegiven(hist_K4me1_dm6_LP88_long, colname1 = colname1)\n",
    "\n",
    "fly_tss_signal_LP88 <- histograms_signalarea_calculatetss(hist_tss_dm6_LP88_long, colname1 = colname1)\n",
    "yeast_rna_signal_LP88 <- histograms_signalarea_calculatetss(hist_rna_sac3_LP88_long, colname1 = colname1)\n",
    "yeast_tss_signal_LP88 <- histograms_signalarea_calculatetss(hist_tss_sac3_LP88_long, colname1 = colname1)\n",
    "#yeast_K4me1_signal_LP88 <- histograms_signalarea_calculatetss(hist_tss_sac3_LP88_long, colname1 = colname1)\n",
    "\n",
    "fly_rna_signal_LP91 <- histograms_signalarea_calculate_sizegiven(hist_rna_dm6_LP91_long, colname1 = colname1)\n",
    "fly_tss_signal_LP91 <- histograms_signalarea_calculatetss(hist_tss_dm6_LP91_long, colname1 = colname1)\n",
    "fly_K36me3_signal_LP91 <- histograms_signalarea_calculate_sizegiven(hist_K36me3_dm6_LP91_long, colname1 = colname1)\n",
    "fly_K4me1_signal_LP91 <- histograms_signalarea_calculate_sizegiven(hist_K4me1_dm6_LP91_long, colname1 = colname1)\n",
    "yeast_rna_signal_LP91 <- histograms_signalarea_calculatetss(hist_rna_sac3_LP91_long, colname1 = colname1)\n",
    "yeast_tss_signal_LP91 <- histograms_signalarea_calculatetss(hist_tss_sac3_LP91_long, colname1 = colname1)\n",
    "\n",
    "fly_tss_signal_LP92 <- histograms_signalarea_calculatetss(hist_tss_dm6_LP92_long, colname1 = colname1)\n",
    "yeast_tss_signal_LP92 <- histograms_signalarea_calculatetss(hist_tss_sac3_LP92_long, colname1 = colname1)\n",
    "\n",
    "fly_rna_signal_LP95 <- histograms_signalarea_calculate_sizegiven(hist_rna_dm6_LP95_long, colname1 = colname1)\n",
    "fly_tss_signal_LP95 <- histograms_signalarea_calculatetss(hist_tss_dm6_LP95_long, colname1 = colname1)\n",
    "yeast_rna_signal_LP95 <- histograms_signalarea_calculatetss(hist_rna_sac3_LP95_long, colname1 = colname1)\n",
    "yeast_tss_signal_LP95 <- histograms_signalarea_calculatetss(hist_tss_sac3_LP95_long, colname1 = colname1)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "de1daaf0-2b9d-4dea-835d-a21d04dde14e",
   "metadata": {},
   "outputs": [],
   "source": [
    "head(fly_K4me1_signal_LP81 %>% filter(grepl(\"Hela\",ID)) , n = 30)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "29476f53-c987-4614-9f4d-43772a45b35b",
   "metadata": {},
   "outputs": [],
   "source": [
    "fly_tss_signal_plot <- fly_tss_signal\n",
    "fly_tss_signal_plot$antibody = str_match(fly_tss_signal$ID, '([^_]+)(?:_[^_]+){1}$')[,2]\n",
    "fly_tss_signal_plot$timepoint = str_match(fly_tss_signal$ID, '([^_]+)(?:_[^_]+){3}$')[,2]\n",
    "fly_tss_signal_plot$treatment = str_match(fly_tss_signal$ID, '([^_]+)(?:_[^_]+){4}$')[,2]\n",
    "fly_tss_signal_plot$biorep = str_match(fly_tss_signal$ID, '([^_]+)(?:_[^_]+){2}$')[,2]\n",
    "head(fly_tss_signal_plot)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "853d8cd1-3e96-49ec-be4d-1b107111c505",
   "metadata": {},
   "outputs": [],
   "source": [
    "fly_peak_K27ac_signal$ID <- gsub(\"\\\\.Coverage\", \"\", as.character(fly_peak_K27ac_signal$ID))\n",
    "colnames(fly_peak_K27ac_signal)[1] <- \"fly_K27ac_signal\"\n",
    "\n",
    "fly_peak_K4me3_signal$ID <- gsub(\"\\\\.Coverage\", \"\", as.character(fly_peak_K4me3_signal$ID))\n",
    "colnames(fly_peak_K4me3_signal)[1] <- \"fly_K4me3_signal\"\n",
    "\n",
    "fly_peak_K36me3_signal$ID <- gsub(\"\\\\.Coverage\", \"\", as.character(fly_peak_K36me3_signal$ID))\n",
    "colnames(fly_peak_K36me3_signal)[1] <- \"fly_K36me3_signal\"\n",
    "\n",
    "fly_tss_signal$ID <- gsub(\"\\\\.Coverage\", \"\", as.character(fly_tss_signal$ID))\n",
    "colnames(fly_tss_signal)[1] <- \"fly_tss_signal\"\n",
    "\n",
    "yeast_tss_signal$ID <- gsub(\"\\\\.Coverage\", \"\", as.character(yeast_tss_signal$ID))\n",
    "colnames(yeast_tss_signal)[1] <- \"yeast_tss_signal\"\n",
    "#head(fly_tss_IPs_signal)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "16c9961f-3d64-4a51-b956-c8382224da85",
   "metadata": {},
   "source": [
    "## spike-in normalized plots"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "2557855f-6742-44fd-a4ff-00bebd0a4bf1",
   "metadata": {},
   "outputs": [],
   "source": [
    "genecov_all_norm_sep_method_LP88 <- genecov_all_norm_LP88 %>% pivot_wider(\n",
    "  names_from = method, values_from = Coverage\n",
    ")\n",
    "\n",
    "genecov_all_norm_sep_method_LP88 <- genecov_all_norm_sep_method_LP88 %>%\n",
    "  mutate(confidence = abs(flynorm - yeastnorm)) %>%\n",
    "mutate(dualnorm = (flynorm + yeastnorm)/2)\n",
    "\n",
    "genecov_all_norm_sep_method_LP88_long <- genecov_all_norm_sep_method_LP88 %>%\n",
    "  pivot_longer(cols = c(\"flynorm\", \"yeastnorm\"), \n",
    "               names_to = \"spike.in\", \n",
    "               values_to = \"Normalized.Coverage\") "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "43d1d622-7b66-443e-a14b-0b74115ee074",
   "metadata": {},
   "outputs": [],
   "source": [
    "plot_histograms_stdcurve <- function(input_df, brewer_cols) {\n",
    "  xvar <- colnames(input_df)[1]\n",
    "  condition_num <- length(unique(input_df[[5]]))\n",
    "  antibody_name <- input_df[[6]]\n",
    "  cell_name <- input_df[[3]]\n",
    "  brewer_pallet <- as.numeric(brewer_cols)\n",
    "\n",
    "  input_df |>\n",
    "    ggplot2::ggplot(ggplot2::aes(x = Distance_from_tss, y = dualnorm, \n",
    "                                 group=Sample, color = as.factor(as.numeric(str_remove(timepoint, \"inter\"))))) +\n",
    "    ggplot2::geom_line(alpha = 0.9, linewidth = 1.1) +\n",
    "    ggplot2::labs(title = paste(\"Histogram of\", antibody_name, \"in\", cell_name),\n",
    "         x = xvar,\n",
    "         y = colnames(input_df[8])) +\n",
    "    ggplot2::scale_color_manual(\n",
    "      values = colorRampPalette(\n",
    "        RColorBrewer::brewer.pal(brewer_pallet, \"Blues\"))(condition_num+4)[(4):(condition_num+4)],\n",
    "      name = \"Condition\",\n",
    "      labels = as.factor(as.numeric(str_remove(unique(input_df[[5]]), \"inter\")))) +\n",
    "    ggplot2::theme_classic() +\n",
    "    ggplot2::theme(legend.position = c(0.84, 0.76),\n",
    "          legend.background = element_rect(\n",
    "            size=0.7, linetype=\"solid\",\n",
    "            colour =\"grey20\"))\n",
    "}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "5390ed71-4842-4942-84b3-afcab6e76412",
   "metadata": {},
   "outputs": [],
   "source": [
    "plot_histograms_dualnorm_natgen <- function(input_df, treatment_select, biorep_select, antibody_select, regions) {\n",
    "  \n",
    "  # extract y column name\n",
    "  yvar <- \"Normalized.Coverage\"\n",
    "  \n",
    "  ggplot(input_df %>%\n",
    "           filter(antibody == antibody_select) %>%\n",
    "           filter(biorep == biorep_select) %>%\n",
    "           filter(treatment %in% c(\"DMSO\", treatment_select))) +\n",
    "    aes(x = Distance_from_tss,\n",
    "        group = interaction(treatment, timepoint, biorep, techrep, spike.in),\n",
    "        color = timepoint) + \n",
    "    geom_line(aes(y = .data[[yvar]], linetype = spike.in)) +\n",
    "    geom_ribbon(aes(ymax = dualnorm + confidence * 0.5,\n",
    "                    ymin = dualnorm - confidence * 0.5,\n",
    "                    fill = timepoint),\n",
    "                alpha = 0.4, color = NA) +\n",
    "    theme(legend.position = c(0.86, 0.75)) +\n",
    "    labs(title = paste(\"Dual spike-in normalized\", antibody_select, \"around\", regions, treatment_select), \n",
    "         subtitle = paste(\"Biological replicate\", biorep_select),\n",
    "         x = paste(\"Distance from\", regions, \"center\"),\n",
    "         y = \"Normalized Coverage\") +\n",
    "    scale_color_manual(values = c(\"firebrick\", \"slateblue\", \"goldenrod2\")) +\n",
    "    scale_fill_manual(values = c(\"firebrick\", \"slateblue\", \"goldenrod2\")) +\n",
    "    scale_linetype_manual(values = c(\"dotted\", \"dashed\"),\n",
    "                          name = \"spike-in species\",\n",
    "                          labels = c(\"fly normalized\", \"yeast normalized\"))\n",
    "}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "c52350de-cab6-4328-9ca6-64cd8f8c57cf",
   "metadata": {},
   "outputs": [],
   "source": [
    "histograms_signalarea_stdcurve <- function(input_df, Sample, dualnorm, colname1) {\n",
    "  # ensym() converts strings into symbols, so function inputs can be understood\n",
    "  ### within ggplot\n",
    "  Sample <- rlang::ensym(Sample)\n",
    "  dualnorm <- rlang::ensym(dualnorm)\n",
    "  colname1 <- rlang::ensym(colname1)\n",
    "  # get a list of unique sample names\n",
    "  samples <- unique(input_df$Sample)\n",
    "\n",
    "  # make sure first column is named properly\n",
    "  colnames(input_df)[1] <- \"Distance_from_center\"\n",
    "\n",
    "  # get x variable\n",
    "  x <- unique(input_df[[1]])\n",
    "\n",
    "  # initialize empty matrix, then use to make dataframe output\n",
    "  AUC_peaks <- matrix(data = \"\", nrow = length(samples), ncol = 1)\n",
    "  AUC_peaks_df <- data.frame(AUC_peaks, row.names = samples)\n",
    "\n",
    "  # fill in AUC dataframe for each sample\n",
    "  for (i in 1:length(samples)) {\n",
    "\n",
    "    y <- input_df |>\n",
    "      dplyr::filter(Sample == samples[i]) |>\n",
    "      dplyr::select(dualnorm)\n",
    "    y <- dplyr::pull(y, dualnorm)\n",
    "\n",
    "    AUC_peaks_df[i, ] <- DescTools::AUC(x, y, method = c(\"trapezoid\"))\n",
    "\n",
    "  }\n",
    " # uncomment below if this stops working\n",
    "  #AUC_peaks_df\n",
    "\n",
    "  # get data in numeric form\n",
    "  AUC_peaks_df <- as.data.frame(sapply(AUC_peaks_df, as.numeric))\n",
    "  #now get this also as an output:\n",
    "  AUC_peaks_df\n",
    "\n",
    "  # add in the treatment column for plotting\n",
    "  treatment_df <- input_df |>\n",
    "    dplyr::filter(Distance_from_center == \"-2000\") |>\n",
    "    dplyr::select(timepoint)\n",
    "\n",
    "  treatment_vect <- dplyr::pull(treatment_df, timepoint)\n",
    "  treatment_vect <- str_remove(treatment_vect, \"inter\")\n",
    " # QC the treatment_vect\n",
    "  print(treatment_vect)\n",
    "\n",
    "  # plot signal area\n",
    "  ## first make colors vector\n",
    "  color_vect <- unique(treatment_vect)\n",
    "\n",
    "  # graph labels in title custom\n",
    "  cell_type <- unique(input_df[[3]])\n",
    "  AUC_name <- colnames(AUC_peaks_df[1])\n",
    "\n",
    "  # plot\n",
    "    AUC_peaks_df |>\n",
    "    ggplot2::ggplot(ggplot2::aes(x = as.numeric(treatment_vect), \n",
    "                                 color = as.factor(as.numeric(treatment_vect)),\n",
    "                                y = AUC_peaks)) +\n",
    "    ggplot2::geom_point(size = 5, alpha = 0.7) +\n",
    "    ggplot2::scale_color_manual(\n",
    "      values = colorRampPalette(RColorBrewer::brewer.pal(9, \"Spectral\"))(6)[1:6],\n",
    "      name = \"Condition\") +\n",
    "    ggplot2::theme_classic() +\n",
    "    ggplot2::labs(title = paste(as.character(AUC_name), \"in\", as.character(cell_type), \"cells\"),\n",
    "         x = \"Condition\",\n",
    "         y = \"Signal Area\")\n",
    "\n",
    "}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "9d9b2fab-a0eb-42c4-963a-a66b19fda711",
   "metadata": {},
   "outputs": [],
   "source": [
    "options(repr.plot.width = 9, repr.plot.height = 7)\n",
    "\n",
    "histograms_signalarea_stdcurve(genecov_all_norm_sep_method, colname1 = colname1) + \n",
    "  labs(title = \"K9ac around Peaks\", \n",
    "       subtitle = \"Read normalized\")"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "chip-wrangler-Renv",
   "language": "R",
   "name": "chip-wrangler-renv"
  },
  "language_info": {
   "codemirror_mode": "r",
   "file_extension": ".r",
   "mimetype": "text/x-r-source",
   "name": "R",
   "pygments_lexer": "r",
   "version": "4.4.2"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
